### PR TITLE
Cache as read layer for the client

### PR DESCRIPTION
**Pull request by bartfish**
## Problem
Now, whenever there was a request made to the API, the sidechain was queried. It had very bad performance, because many connections were made at a time to read data from the contract.

## Solution
This PR introduces caching layer. The client now can only read from cache (so the response is given almost instantly & number of clients calling the api has no impact on the sidechain). 

- [x] cache layer - save/get/flush/delete
- [x] function for automatic cache update

Cache is updated every hour (interval can be set to a different one if needed). ~~To control what happens in the cache there are 2 new endpoints introduced with this PR:~~
- [x] ~~`/clearCache` - to clear the cache if necessary~~ - removed
- [x]  ~~`/updateCache` - if we need results from the sidechain at the very moment, we can call this function and force cache to update, without a need to wait until the time of automatic update~~ - removed

### CODE REVIEW

**Comment by rinor** [comment]
**Note**
- clearCache
- updateCache

are to be considered administrative/sensitive operations/apis so they need to be protected behind an api auth key (in this case since running on heroku) and not left open to public since can be easily abused and force the service to query the sidechain over and over.

**Review by nicarq — CHANGES_REQUESTED** [review]
Cache can't be an action that's updated through public endpoints. Adding auth keys would add unnecessary complexity. This service should be fairly simple and cache should be updated by just restarting the node.

**Comment by bartfish** [comment]
Thanks @nicarq and @rinor! Yeah, if we wanted such endpoints they should have restricted access only. I also wasn't sure how complex we want to go and if restarting the service is enough for `maintenance`. So if it is, then indeed those endpoints don't have to be there - I removed them. Anyway, if we feel like it's something useful to have later, we have the code for this in the `CacheManager` class, so it'll be only required to create endpoints themselves.

File: README.md [hunk 1]
```diff
@@ -44,61 +49,67 @@ heroku config:set USE_NPM_INSTALL=true -a allowedlist-service
...
 <details>
     <summary>/v1/stargate</summary>
-    Returns stargate address, tll_expiry and list of assets.
+    Returns stargate address, tll_expire, cache_update and list of assets.
     (For now only stargate address is fetched from bridge contract)
 
 ```json
 {
     "current_address": "addr1...",
-    "ttl_expiry": 123,
```
Comment (SebastienGllmt) [inline README.md:101]: Changing fields will cause issues in Flint. Please don't do this

File: src/index.ts [hunk 2]
```diff
@@ -40,59 +61,63 @@ const isAddressAllowed = async (req: Request, res: Response) => {
...
-            },
-            assets: tokenRegistry.assets,
-        });
+
+    try {
+        const stargateAddress = await cacheManager.get(CacheKeys.STARGATE) as string;
+        const tokenRegistry = await cacheManager.get(CacheKeys.TOKEN_REGISTRY) as TokensRegistry;
+        
+        // TODO: Update config so node parses this env variable as a Boolean
+        // TODO: do we still need to distinguish between mainnet & devnet here?
```
Comment (SebastienGllmt) [inline src/index.ts:91]: No need to differentiate

File: src/index.ts [hunk 3]
```diff
@@ -17,18 +19,37 @@ declare const CONFIG: ConfigType;
...
-    const allowList = await contract.getAccountsList();
-    if (allowList instanceof Error) {
-        res.status(400).send({ error: allowList.message });
+    try {
+        const allowList = await cacheManager.get(CacheKeys.FULL_ALLOWED_LIST) as string;
+        res.status(200).send({ allowList });
+    } catch (e) {
+        const err = e as Error;
+        console.log(`${err.name}, ${err.message}, ${err.stack}`);
+        res.status(400).send({ error: `Couldn't fetch nnformation about allowed list. ${err.message}` });
```
Comment (SebastienGllmt) [inline src/index.ts:50]: nit: typo

File: src/index.ts (hunk 2 above)
Comment (bartfish) [reply src/index.ts:91]: Oh perfect, I wasn't sure about this one. Thanks! I'll remove that

File: src/cache/index.ts [hunk 4]
```diff
@@ -0,0 +1,65 @@
...
+        for (let action of actions) {
+            console.log(`Updating ${action.key}`);
+            const result = await action.method();
+            // console.log(result);
+            this.save(action.key, result);
+        }
+    }
+
+    public keepCached = async (actions: CacheOption[], intervalMs: number = 20000): Promise<void> => {
+        await this.updateCache(actions);
```
Comment (SebastienGllmt) [inline src/cache/index.ts:38]: You can replace this with [setInterval](https://developer.mozilla.org/en-US/docs/Web/API/setInterval)

Comment (bartfish) [reply src/cache/index.ts:38]: Sure, I was thinking of using it but `setInterval` has one unsafe feature - if something goes wrong inside the interval, another call is scheduled no matter what happened. Although as I think about it now, that's actually a feature here - if we don't get response from the node, we are sure function will try fetch data again...

I agree that in this case setInterval is also good to use, so I'll change it. Thanks!

Comment (bartfish) [reply src/cache/index.ts]: @SebastienGllmt done in the latest commit

**Review by NicolasDP — APPROVED** [review]

**Pull request merged**